src/index.ts

```diff
@@ -10,22 +30,354 @@ import {list} from 'white-space-x';
+}
+
+export function normalizeSpaces10(string: string): string {
+  const chars = Buffer.from(string);
+  const processedChars = Buffer.alloc(chars.length, SPACE_CODE);
+
+  let lastProcessedIndex = -1;
+  let isLastCharWhitespace = 0;
+
+  for (let i = 0; i < chars.length; i++) {
+    const char = chars[i];
+    const isWhitespace = isWhiteSpaceCharCodeNumber(char);
+
+    if (!isWhitespace) {
+      lastProcessedIndex++;
+
+      processedChars[lastProcessedIndex] = char;
+    } else {
+      if (isLastCharWhitespace) {
+        continue;
+      }
+
+      lastProcessedIndex++;
+    }
+
+    isLastCharWhitespace = isWhitespace;
+  }
+
+  // If the original string has whitespaces at the start or at the end, the first and the
+  // last processed chars should be single whitespaces, so we need to make offset from both
+  // ends to trim they
+  const firstChar = processedChars[0];
+  const lastChar = processedChars[lastProcessedIndex];
+
+  const startIndex = firstChar === SPACE_CODE ? 1 : 0;
+  const endIndex = lastChar === SPACE_CODE ? lastProcessedIndex : lastProcessedIndex + 1;
+
+  return processedChars.toString('utf-8', startIndex, endIndex);
 }
 
-function replaceAllWhitespaces(string: string): string {
-  let finalString = '';
+export function normalizeSpaces11(string: string): string {
+  const chars = Buffer.from(string);
+  const processedChars = Buffer.alloc(chars.length, SPACE_CODE);
+
+  let lastProcessedIndex = -1;
+  let isLastCharWhitespace = 0;
+
+  for (let i = 0; i < chars.length; i++) {
+    const char = chars[i];
+    const isWhitespace = isWhiteSpaceCharCodeIndexedUint8Array(char);
+
+    if (!isWhitespace) {
+      lastProcessedIndex++;
 
-  for (const item of list) {
-    const firstIndex = string.indexOf(item.string);
-    const lastIndex = string.lastIndexOf(item.string);
+      processedChars[lastProcessedIndex] = char;
+    } else {
+      if (isLastCharWhitespace) {
+        continue;
+      }
 
-    if (firstIndex === lastIndex) {
-      continue;
+      lastProcessedIndex++;
     }
 
-    finalString = `${string.substring(0, firstIndex)} ${string.substring(lastIndex + 1)}`;
+    isLastCharWhitespace = isWhitespace;
+  }
+
+  // If the original string has whitespaces at the start or at the end, the first and the
+  // last processed chars should be single whitespaces, so we need to make offset from both
+  // ends to trim they
+  const firstChar = processedChars[0];
+  const lastChar = processedChars[lastProcessedIndex];
+
+  const startIndex = firstChar === SPACE_CODE ? 1 : 0;
+  const endIndex = lastChar === SPACE_CODE ? lastProcessedIndex : lastProcessedIndex + 1;
+
+  return processedChars.toString('utf-8', startIndex, endIndex);
+}
+
+export function normalizeSpaces12(filePath: string): string {
+  const res = execSync(
+    `cat ${filePath} | tr -s "[:space:]" | tr "[:space:][:space:]" " " | tr -s "[:space:]"`,
+    {maxBuffer: 999999999999}
+  );
```

Comment:
Interesting solution) Does lambda allow to exec external processes? And it is dependent on OS.

Reply:
@batovpasha yes, it allows! And `tr` is available in the lambda runtime
<img width="380" alt="CleanShot 2022-02-08 at 12 42 05@2x" src="https://user-images.githubusercontent.com/3817380/152980489-0ecd3bd4-ff71-44e0-a198-12b3301d6e05.png">